### ISSUE TITLE
Not mapping proxy classes from doctrine lazy load

### ISSUE THREAD

**Issue by yhnehc** [mark-gerarts/automapper-plus#9]
When you do lazy loading sometimes the objects come back as a proxy class which breaks since there's no mapping config registered for the proxy classes.

Example of the error:
No mapping registered for converting an instance of class Proxies\CG\Bundle\Entity\ProductItem into one of \ProductItemDto

**Comment by mark-gerarts** [comment]
Could you provide some sample code for this? That would help a lot to figure out this issue!

**Comment by dingledorf** [comment]
He's referring to whats happening in this article:

https://www.uvdesk.com/en/blog/doctrine-proxy-objects-lazy-loading/.

The current work around is to use fetch EAGER, I'm not sure if you guys can do anything.

**Comment by mark-gerarts** [comment]
Have you tried calling the getter explicitly?

```
$config
  ->registerMapping(...)
  ->forMember('doctrineProperty', function (YourSourceClass $source) {
      return $source->getDoctrineProperty();
  });
```

This should create a proper instance out of the proxy object, if I'm not mistaken.

**Comment by dingledorf** [comment]
I think the problem is that these proxy classes appear to be dynamically made by Doctrine so you have no way to register the mapping.

**Comment by ErikJson** [comment]
Perhaps a solution could be to, optionally, allow for AutoMapper to look at the parent class of `$from` for a matching mapping?

**Comment by mark-gerarts** [comment]
Alright, I understand the situation now. @ErikJson I think looking at the parent class wouldn't help, because the proxy object still won't be initialized. The exception would no longer be thrown, but there wouldn't be any data to map (correct me if I'm wrong).

One solution could be to provide a way to 'prepare' the source object. Something like this for example:

```php
$config->registerMapping(Source::class, Destination::class)
    ->doBefore(function (Source $source) {
        // Ensure the related entity is loaded by Doctrine.
        $source->getRelatedEntity();
    })
    ->forMember('relatedEntity', Operation::mapTo(...);
```
Then we could also provide this behaviour globally using the `Options`, so we can avoid duplication. 

```php
$options->setBeforeAction(function ($source) {
    // Ensure Proxy objects are loaded.
    if ($source instanceof Proxy) {
        $source->__load();
    }
});
```

The second code snippet would still need a way to circumvent the UnregisteredMappingException for Proxy classes though. 

Any thoughts on this? Or some alternatives?

**Comment by wgnd93** [comment]
I do the following:

`use DoctrineProxies\__CG__\App\Entities\MyEntity as MyEntityDoctrine;

$config->registerMapping(MyEntity::class, MyEntityDTO::class)
->forMember("myEntityProperty", Operation:mapTo(MyEntityDTO::class));

$config->registerMapping(MyEntityDoctrine::class, MyEntityDTO::class)
->copyFrom(MyEntity::class, MyEntityDTO::class);`

**Comment by wgnd93** [comment]
I do the following:
`php  
use DoctrineProxies_CG_\App\Entities\MyEntity as MyEntityDoctrine;

$config->registerMapping(MyEntity::class, MyEntityDTO::class)
->forMember("myEntityProperty", Operation:mapTo(MyEntityDTO::class));

$config->registerMapping(MyEntityDoctrine::class, MyEntityDTO::class)
->copyFrom(MyEntity::class, MyEntityDTO::class);

php`

**Comment by wgnd93** [comment]
I do the following:

```php use DoctrineProxies\__CG__\App\Entities\MyEntity as MyEntityDoctrine;

$config->registerMapping(MyEntity::class, MyEntityDTO::class)
->forMember("myEntityProperty", Operation:mapTo(MyEntityDTO::class));

$config->registerMapping(MyEntityDoctrine::class, MyEntityDTO::class)
->copyFrom(MyEntity::class, MyEntityDTO::class);```

**Comment by wgnd93** [comment]
I do the following:

```php use DoctrineProxies\__CG__\App\Entities\MyEntity as MyEntityDoctrine;

$config->registerMapping(MyEntity::class, MyEntityDTO::class)
->forMember("myEntityProperty", Operation:mapTo(MyEntityDTO::class));

$config->registerMapping(MyEntityDoctrine::class, MyEntityDTO::class)
->copyFrom(MyEntity::class, MyEntityDTO::class);

**Comment by mschop** [comment]
@mark-gerarts I've added a suggestion for resolving this issue in PR #21 .

**Comment by mark-gerarts** [comment]
Reopening because I want to test if #21 actually fixes this issue

**Comment by r-alexandr** [comment]
I have the same problem with proxy classes of doctrine.
Object A has a reference to object B. I get object A without loading B.  I map object A into Adto1. Adto1 has no related objects. It's all good. Then I get object A with forced load B. I map object A into Adto2. Adto2 has an associated Bdto object. I get an error: 
"No mapping registered for converting an instance of class Proxies\\__CG__\\App\\B into one of App\\Bdto"

```
calss A
{
	public id int;
	public name string;
	public b B;
}
calss Adto1
{
	public id int;
	public name string;
}
calss Adto2
{
	public id int;
	public name string;
	public b Bdto;
}

calss B
{
	public id int;
	public name string;
}

calss Bdto
{
	public id int;
	public name string;
}

$config->registerMapping(A::class, Adto1::class);
$config->registerMapping(B::class, Bdto::class);
$config->registerMapping(A::class, Adto2::class)
	->forMember('b', Operation::mapTo(Bdto::class));

   
$entityA = $repository->find(100);
$mapper->map($entityA, Adto1::class);

$entityA1 = $repository->createQueryBuilder('a')
            ->innerJoin('a.b','b')
            ->getQuery()
            ->getOneOrNullResult();

$mapper->map($entityA1, Adto2::class);

```

**Comment by mark-gerarts** [comment]
Hi @r-alexandr. I've tested your example, and the issue seems fixed with PR #21. I've tagged a new release containing this fix (1.2.2), so if you run a `composer update mark-gerarts/auto-mapper-plus` the exception should disappear. If you still run into problems, feel free to reopen.
